src/OrbitQt/AnnotatingSourceCodeHandler.cpp

```diff
@@ -0,0 +1,178 @@
+  local_source_file_path_ = std::filesystem::path{location_info_.source_file()};
+
+  if (!std::filesystem::exists(local_source_file_path_)) {
+    orbit_source_paths_mapping::MappingManager mapping_manager{};
+    std::optional<std::filesystem::path> maybe_mapping_file_path =
+        mapping_manager.MapToFirstExistingTarget(local_source_file_path_);
+    if (!maybe_mapping_file_path.has_value()) {
+      dialog_->SetStatusMessage(
+          QString("Could not find the source code file \"%1\" on this machine.")
+              .arg(QString::fromStdString(local_source_file_path_.string())),
+          "Choose file...");
+      awaited_button_action_ = ButtonAction::kChooseFile;
+      return;
+    }
+
+    local_source_file_path_ = std::move(maybe_mapping_file_path.value());
+  }
+
+  ErrorMessageOr<std::string> source_file_contents_or_error =
+      orbit_base::ReadFileToString(local_source_file_path_);
+
+  if (source_file_contents_or_error.has_error()) {
+    dialog_->SetStatusMessage(
+        QString("Error while reading source code file \"%1\":\n%2")
+            .arg(QString::fromStdString(local_source_file_path_.string()),
+                 QString::fromStdString(source_file_contents_or_error.error().message())),
+        "Choose another file...");
+    awaited_button_action_ = ButtonAction::kChooseFile;
+    return;
+  }
+
+  HandleSourceCode(QString::fromStdString(source_file_contents_or_error.value()));
+}
+
+void AnnotatingSourceCodeHandler::ChooseFile() {
+  if (dialog_.isNull()) return;
+
+  ErrorMessageOr<std::optional<QString>> result =
+      orbit_source_paths_mapping_ui::ShowFileOpenDialogAndReadSourceFile(
+          dialog_, local_source_file_path_, true);
+
+  if (result.has_error()) {
+    dialog_->SetStatusMessage(QString::fromStdString(result.error().message()),
+                              "Choose another file...");
+    return;
+  }
+
+  if (result.value() == std::nullopt) {
+    // The user aborted the operation. We change nothing in that case.
+    return;
+  }
+
+  HandleSourceCode(result.value().value());
+}
+
+void AnnotatingSourceCodeHandler::HandleSourceCode(const QString& source_file_contents) {
+  annotations_ = orbit_code_report::AnnotateDisassemblyWithSourceCode(
+      function_info_, location_info_, source_file_contents.toStdString(), elf_file_.get(), report_);
+
+  // When loading the source code takes less than the given time we won't ask the user if they want
+  // to have the source code added for context. We will add it right away - as it's done in Visual
+  // Studio.
+  constexpr std::chrono::steady_clock::duration kMaxWaitingTime = std::chrono::milliseconds{250};
+
+  // ButtonAction::kChooseFile means the user had to select the source file path manually. In this
+  // case we don't keep them waiting even if loading took longer than the limit.
+  if (std::chrono::steady_clock::now() - starting_time_ < kMaxWaitingTime ||
+      awaited_button_action_ == ButtonAction::kChooseFile) {
```

Comment:
pers pref: in combination with how the comment is worded
`awaited_button_action_ == ButtonAction::kChooseFile || std::chrono::steady_clock::now() - starting_time_ < kMaxWaitingTime`

Reply:
I agree. Done.